### DISCUSSION TITLE
The MFA Feature doesn't work

### DISCUSSION THREAD

**Discussion by ar3diu** [reflexsoar/reflex-api#197] [Q&A] [answered]
I can't activate MFA for a ReflexSOAR user account.

![image](https://user-images.githubusercontent.com/61625853/174313835-06f7980d-0f93-404a-bf44-07a81ef9e240.png)

Although if I enforce MFA for a user in the User Settings page, she will be prompted with the MFA page at the next logon but without giving her the option to set up the MFA in advance.

![image](https://user-images.githubusercontent.com/61625853/174355471-f0cbdcbd-b2a8-4c50-9fa7-8d108c6ab5c7.png)

![image](https://user-images.githubusercontent.com/61625853/174314696-a43706d7-ab34-4ada-ab2c-c2586d868171.png)

 Not sure if this feature should have worked in version 2022-05-01 but it's worth asking.

**Comment by n3tsurge** [comment]
This feature should be working as of 2022.05.01 however you can't "force" MFA on for a user yet, which is what the UI element you used is intended to do, it is in our backlog to allow "forcing" MFA.  The workflow for the user to initially set up MFA when it is forced on them does not exist yet.

Instead users have to go to the Profile page and click Enable MFA

![image](https://user-images.githubusercontent.com/1574741/174315936-b18a3e12-4c11-4074-a534-39bc1aca3ac8.png)

**Reply by ar3diu** [reply]
I understand. However, the first screenshot is the result of my attempt to enable MFA by following the exact same steps you suggested.

**Reply by n3tsurge** [reply]
Okay that's weird.  I just walked through the MFA setup on my side.  Which authenticator are you using?  If you let your code cycle twice before trying the code does that make a difference?

**Reply by ar3diu** [reply]
I let several codes cycle without success though.
I tried with both [Google Authenticator](https://play.google.com/store/apps/details?id=com.google.android.apps.authenticator2&hl=ro&gl=US) and [Aegis](https://play.google.com/store/apps/details?id=com.beemdevelopment.aegis&hl=ro&gl=US) on a phone with Android 12.

**Reply by n3tsurge** [reply]
Can you provide me (obfuscated of course) the output of the following queries via the Elastic API.  I want to make sure the user is actually getting an otp_secret and that the mfa_enabled field is actually being set.  I also want to make sure the index settings are correct

```
GET reflex-users/_search
{
  "query": {
    "exists": {
      "field": "otp_secret"
    }
  }
}
```

```
GET reflex-users/_settings
```

**Reply by n3tsurge** [reply]
Can you also supply the output of 

```
GET reflex-settings/_search
{}
```

**Reply by ar3diu** [reply]
Let me know if I "redacted" too much :smile: 

```
{
  "_index" : "reflex-users-0.1.4",
  "_id" : "redacted",
  "_score" : 1.0,
  "_source" : {
    "organization" : "redacted",
    "username" : "Test",
    "first_name" : "John",
    "last_name" : "Doe",
    "email" : "user@reflexsoar.com",
    "mfa_enabled" : true,
    "password_hash" : "redacted",
    "created_at" : "2022-06-17T16:05:46.340351",
    "created_by" : {
      "username" : "admin",
      "uuid" : "5343b529-e55f-4f45-a84b-7ab0734745cb",
      "organization" : "redacted"
    },
    "uuid" : "72d2dedf-0c4d-49c9-86d7-bcc1cee0ccbf",
    "updated_at" : "2022-06-17T20:45:40.440234",
    "updated_by" : {
      "username" : "Admin",
      "uuid" : "efcfd178-6c56-41cd-98bc-bd2848df1403",
      "organization" : "redacted"
    },
    "deleted" : false,
    "failed_logons" : 0,
    "last_logon" : "2022-06-17T20:23:47.817684",
    "locked" : false,
    "otp_secret" : "redacted",
    "default_org" : true,
    "role_uuid" : "5c152daf-48ff-4901-bfae-0f01003de564"
  }
}
```

```
{
  "reflex-users-0.1.4" : {
    "settings" : {
      "index" : {
        "routing" : {
          "allocation" : {
            "include" : {
              "_tier_preference" : "data_content"
            }
          }
        },
        "refresh_interval" : "1s",
        "number_of_shards" : "1",
        "provided_name" : "reflex-users-0.1.4",
        "creation_date" : "1655282940928",
        "number_of_replicas" : "1",
        "uuid" : "redacted",
        "version" : {
          "created" : "8020099"
        }
      }
    }
  }
}
```


```
{
  "took" : 1,
  "timed_out" : false,
  "_shards" : {
    "total" : 1,
    "successful" : 1,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : {
      "value" : 2,
      "relation" : "eq"
    },
    "max_score" : 1.0,
    "hits" : [
      {
        "_index" : "reflex-settings-0.1.4",
        "_id" : "redacted",
        "_score" : 1.0,
        "_source" : {
          "base_url" : "localhost",
          "required_case_templates" : true,
          "allow_comment_deletion" : false,
          "playbook_action_timeout" : 300,
          "playbook_timeout" : 3600,
          "logon_password_attempts" : redacted,
          "api_key_valid_days" : redacted,
          "agent_pairing_token_valid_minutes" : 15,
          "require_event_dismiss_comment" : false,
          "allow_event_deletion" : false,
          "require_case_close_comment" : false,
          "assign_case_on_create" : true,
          "assign_task_on_start" : true,
          "allow_comment_editing" : false,
          "events_page_refresh" : 60,
          "events_per_page" : 10,
          "require_approved_ips" : false,
          "data_types" : [
            "ip",
            "user",
            "host",
            "fqdn",
            "sha1",
            "md5",
            "sha256",
            "imphash",
            "ssdeep",
            "vthash",
            "network",
            "domain",
            "url",
            "mail",
            "sid",
            "mac"
          ],
          "organization" : "redacted",
          "case_sla_days" : 14,
          "event_sla_minutes" : 5,
          "require_mfa" : false,
          "minimum_password_length" : redacted,
          "enforce_password_complexity" : redacted,
          "disallowed_password_keywords" : [
            "reflex"
          ],
          "created_at" : "2022-06-15T10:51:21.378446",
          "created_by" : {
            "username" : "Admin",
            "uuid" : "redacted",
            "organization" : "redacted"
          },
          "uuid" : "redacted",
          "updated_at" : "2022-06-15T10:51:21.380477",
          "updated_by" : {
            "username" : "Admin",
            "uuid" : "redacted",
            "organization" : "redacted"
          }
        }
      },
      {
        "_index" : "reflex-settings-0.1.4",
        "_id" : "redacted",
        "_score" : 1.0,
        "_source" : {
          "base_url" : "localhost",
          "required_case_templates" : true,
          "allow_comment_deletion" : false,
          "playbook_action_timeout" : 300,
          "playbook_timeout" : 3600,
          "logon_password_attempts" : redacted,
          "api_key_valid_days" : redacted,
          "agent_pairing_token_valid_minutes" : 15,
          "require_event_dismiss_comment" : true,
          "allow_event_deletion" : false,
          "require_case_close_comment" : true,
          "assign_case_on_create" : true,
          "assign_task_on_start" : true,
          "allow_comment_editing" : false,
          "events_page_refresh" : 60,
          "events_per_page" : "100",
          "require_approved_ips" : false,
          "data_types" : [
            "ip",
            "user",
            "host",
            "fqdn",
            "sha1",
            "md5",
            "sha256",
            "imphash",
            "ssdeep",
            "vthash",
            "network",
            "domain",
            "url",
            "mail",
            "sid",
            "mac"
          ],
          "organization" : "redacted",
          "case_sla_days" : 14,
          "event_sla_minutes" : 5,
          "require_mfa" : false,
          "minimum_password_length" : redacted,
          "enforce_password_complexity" : redacted,
          "disallowed_password_keywords" : [
            "reflex"
          ],
          "created_at" : "2022-06-15T10:47:28.192705",
          "uuid" : "e196e0b5-181c-4568-9342-20ba128e094d",
          "updated_at" : "2022-06-15T10:47:28.192806",
          "peristent_pairing_token" : "redacted",
          "email_server" : null,
          "persistent_pairing_token" : null,
          "email_secret_uuid" : null,
          "approved_ips" : null,
          "require_case_templates" : null,
          "email_from" : null
        }
      }
    ]
  }
}
```

**Reply by n3tsurge** [reply]
The only thing that sticks out is that the user is missing the `mfa_enabled` field so I am not sure how it's getting prompted in the first place.  Can you do the following then try enabling it again via logging in to the user and setting it as on in the profile page one last time

```
POST /<reflex_users_index>/_update/<id>
{
    "doc" : {
        "mfa_enabled": "false",
        "otp_secret": null,
    }
}
```

**Reply by ar3diu** [reply]
Oh, my bad!
I removed the user "Test" after I posted the screenshots in the first comment. I forgot about that when I ran the commands in Kibana.
Anyway, I re-enabled the user and ran the first command. I edited the comment so I can update it with the new results. You can see now that MFA was enabled as it was enforced through the User Settings Page.

Then I disabled the MFA by running the last command you posted, tried to manually enable MFA through the Profile page of the user "Test" and got the same error as before: `Invalid MFA Code. Try again`

**Reply by n3tsurge** [reply]
So this is going to seem extremely nuts, can you stop the reflex-api, restart it then try the steps again?

**Reply by n3tsurge** [reply]
As well can you open Dev Tools before performing the request and share the response for the API call to `/api/v2.0/user/validate_mfa_setup`

**Reply by ar3diu** [reply]
Hmm, if I would run that in Kibana Dev Tools, it will run against my Elasticsearch cluster API which I don't think is what you meant.
However, I tried to run it this way: `curl -X GET https://my-reflexsoar-api-host/api/v2.0/user/validate_mfa_setup -u user@reflexsoar.com` and I got `{"message": "Invalid access token."}`. But again, not sure if this is what you meant.

**Reply by n3tsurge** [reply]
Sorry I meant your browsers Dev Tools and pulling the response to that HTTP request in the Network log as you walk through the MFA setup process. I can't replicate the issue locally but I have a suspicion and this will help rule it in or out

**Reply by ar3diu** [reply]
This is what I get when I reach to this step:
![image](https://user-images.githubusercontent.com/61625853/174440013-d655f665-1d0e-471e-864b-0d601b1f2ba2.png)

Response Header:
```
HTTP/1.1 400 BAD REQUEST
Server: nginx/1.21.6
Date: Sat, 18 Jun 2022 14:24:35 GMT
Content-Type: application/json
Content-Length: 34
Connection: keep-alive
Access-Control-Allow-Origin: https://my_soar_host
Vary: Origin
```

Response:
```
{"message": "Invalid TOTP Token"}
```

**Reply by n3tsurge** [reply]
Okay perfect this aligns with the section of code I was looking at.  Has your API container been up for an extended period of time or did you try this right after a container restart? If not can you try to restart the container and immediately do the MFA setup, I believe there's a time issue with the MFA challenge token and restarting the API helps rule it in or out.

Sorry for the 100 questions

**Reply by ar3diu** [reply]
It has been up for a few days. However, I restarted reflex-api and tried to set up MFA again without success.

**Reply by n3tsurge** [reply]
Alright well I know where it's giving you fits now I'll do my best to replicate and figure it out for you.  I appreciate your patience, feedback and interest in the tool.

**Reply by ar3diu** [reply]
Thank you very much, Brian! I'm actively testing it these days and may open more similar discussions as I try to set some things (Orgs, Users, Roles, Inputs) and keep encountering some errors. So bear with me, so to speak :smile:

**Reply by n3tsurge** [reply]
Hopefully you won't find any more! I know @SMAPPER and his team are using it all day every day ao hopefully you don't run in to more stuff.

**Reply by ar3diu** [reply]
Right now I'm trying to understand how organizations work, actually how the whole multi-tenancy concept would work.  For instance, I'm having some issues logging in with some user accounts linked to a new organization I just created. Changing the settings for a new organization also throws an `Internal server error`.

Do you plan to update the documentation before the September release?  It would help a lot to better understand each component.

**Reply by ar3diu** [reply]
As an additional note here, I just tried on version 2022-06 with a token authenticator on iOS and had the same result.

**Comment by n3tsurge** [comment]
Yeah we do need to definitely update the documentation We raced to get most of the basic stuff documented but organizations is one that wasn't fully documented. Essentially the default organization acts as a parent that can access the data of all sub organizations so if you think of it like managed service providers they would be the default organization they would add their customers as sub tenants or sub organizations and the customers would work out of their organization and they won't be able to see anybody else's data but the parent organization can act on their behalf as like managed detection and response. For the error 500 are you seeing anything in the API logs that you can share with me when that happens? You're running into some things that we have not yet experienced in our production instance or in development so I'm curious.


Brian Carroll
________________________________

**Comment by n3tsurge** [comment]
I'm still unable to replicate the issue so I haven't released a fix for it yet

Brian Carroll
________________________________

**Comment by ar3diu** [comment, marked as answer]
If everyone comes across this again, check your timezone settings/status. :timer_clock:
